File: Baekjoon/diamond/1557.py
import sys

input = lambda: sys.stdin.readline().rstrip()
mu = [0 for i in range(44722)]


def setup() -> None:
    mu[1] = 1
    for i in range(1, 44722):
        for j in range(2 * i, 44722, i):
            mu[j] -= mu[i]


def get_SFI_count(k: int) -> int:
    count = 0
    i = 1
    while i * i <= k:
        count += mu[i] * (k // (i * i))
        i += 1
    return count


def get_SFI(search: int, start: int, end: int) -> int:
    mid = (start + end) // 2
    sfi_count = get_SFI_count(mid)

    if sfi_count == search:
        mid -= 1
        while get_SFI_count(mid) == search:
            mid -= 1
        return mid + 1
    elif sfi_count > search:
        return get_SFI(search, start, mid - 1)
    else:
        return get_SFI(search, mid + 1, end)


k = int(input())
setup()
print(get_SFI(k, k, k * 2))

File: Baekjoon/diamond/1557.cpp
#include <iostream>
using namespace std;
using ll = long long;

int mu[44722];


void setup() {
    mu[1] = 1; // 뫼비우스 함숫값 전처리
    for (int i = 1 ; i <= 44721 ; i++) {
        for (int j = 2 * i ; j <= 44721 ; j += i) {
            mu[j] -= mu[i];
        }
    }
}

ll get_SFI_count(ll k) { // k까지의 제곱ㄴㄴ수의 개수
    ll count = 0;
    for (ll i = 1 ; i * i <= k; i++) {
        count += mu[i] * (k / (i * i));
    }
    return count;
}

int get_SFI(int search, ll start, ll end) {
    ll mid = (start+end)/2;
    ll sfi_count = get_SFI_count(mid);
    

    if (sfi_count == search) {
        mid--;
        while (get_SFI_count(mid) == search) {
            mid--;
        }
        return mid+1;
    }
    else if (sfi_count > search) return get_SFI(search, start, mid-1);
    else return get_SFI(search, mid+1, end);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    setup();


    ll k;
    cin >> k;

    cout << get_SFI(k, k, k*2);

    return 0;
}